src/lib/mdns/ServicePool.cpp

```diff
@@ -0,0 +1,272 @@
+            }
+        }
+        chip::Platform::MemoryFree(service->mTextEntryies);
+        service->mTextEntryies = nullptr;
+    }
+}
+
+CHIP_ERROR CopyTextEntries(MdnsService * dest, const MdnsService & source)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    if (source.mTextEntrySize)
+    {
+        dest->mTextEntryies = static_cast<TextEntry *>(chip::Platform::MemoryCalloc(source.mTextEntrySize, sizeof(TextEntry)));
+        VerifyOrExit(dest->mTextEntryies != nullptr, error = CHIP_ERROR_NO_MEMORY);
+        dest->mTextEntrySize = source.mTextEntrySize;
+
+        for (size_t i = 0; i < source.mTextEntrySize; i++)
+        {
+            TextEntry & entry = dest->mTextEntryies[i];
+            size_t keySize    = strnlen(source.mTextEntryies[i].mKey, kMdnsTextMaxSize);
+            char * key        = static_cast<char *>(chip::Platform::MemoryAlloc(keySize + 1));
+            uint8_t * data    = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(source.mTextEntryies[i].mDataSize));
+
+            VerifyOrExit(keySize < kMdnsTextMaxSize, error = CHIP_ERROR_INVALID_ARGUMENT);
+            VerifyOrExit(key != nullptr, error = CHIP_ERROR_NO_MEMORY);
+            VerifyOrExit(data != nullptr, error = CHIP_ERROR_NO_MEMORY);
+            strncpy(key, source.mTextEntryies[i].mKey, keySize + 1);
+            entry.mDataSize = source.mTextEntryies[i].mDataSize;
+            memcpy(data, source.mTextEntryies[i].mData, entry.mDataSize);
+            entry.mKey  = key;
+            entry.mData = data;
+        }
+    }
+
+exit:
+    if (error != CHIP_NO_ERROR)
+    {
+        FreeTextEntries(dest);
+    }
+
+    return error;
+}
+
+} // namespace
+
+namespace chip {
+namespace Mdns {
+
+ServicePool::Entry::Entry() : mHasValue(false), mIsTombstone(false), mNodeId(0), mFabricId(0)
+{
+    mService.mTextEntryies  = nullptr;
+    mService.mTextEntrySize = 0;
+}
+
+ServicePool::Entry::Entry(Entry && rhs)
+{
+    MoveFrom(std::move(rhs));
+}
+
+ServicePool::Entry & ServicePool::Entry::MoveFrom(Entry && rhs)
+{
+    mService     = rhs.mService;
+    mNodeId      = rhs.mNodeId;
+    mFabricId    = rhs.mFabricId;
+    mHasValue    = rhs.mHasValue;
+    mIsTombstone = rhs.mIsTombstone;
+
+    rhs.mHasValue               = false;
+    rhs.mIsTombstone            = false;
+    rhs.mService.mTextEntryies  = nullptr;
+    rhs.mService.mTextEntrySize = 0;
+
+    return *this;
+}
+
+CHIP_ERROR ServicePool::Entry::Emplace(const MdnsService & service, uint64_t nodeId, uint64_t fabricId)
+{
+    CHIP_ERROR error;
+
+    if (mHasValue)
+    {
+        FreeTextEntries(&mService);
+    }
+    mService               = service;
+    mNodeId                = nodeId;
+    mFabricId              = fabricId;
+    mService.mTextEntryies = nullptr;
+    error                  = CopyTextEntries(&mService, service);
+
+    if (error == CHIP_NO_ERROR)
+    {
+        mHasValue    = true;
+        mIsTombstone = false;
+    }
+    return error;
```

Comment:
So on failure here, if `mHasValue` used to be true before this call, it remains true, right?  But our mService has null entries and so forth.

Is that OK, or should we set `mHasValue` to false on error?  Previous version of the changes used to do that, by calling `Clear`....